File: backend/app/parser.py
"""Parse Flipkart 'My Orders' list + per-order detail into the frontend Order shape.

Field paths mirror flipkart.com's internal APIs discovered live. An order holds a MAP of
units (line items); we collapse to one row per order, preferring the real product over
free add-ons (VAS: memberships, trust shield) and the most actionable delivery status.
"""

from __future__ import annotations

import re
from datetime import datetime, timezone
from typing import Any

from .errors import parse_error

# --- generic helpers --------------------------------------------------------

def get(obj: Any, path: str) -> Any:
    cur = obj
    for key in path.split("."):
        if not isinstance(cur, dict):
            return None
        cur = cur.get(key)
    return cur


def _num(v: Any) -> float | None:
    try:
        n = float(v)
    except (TypeError, ValueError):
        return None
    return n if n > 0 else None


def _to_iso(ms: Any) -> str | None:
    n = _num(ms)
    if n is None:
        return None
    try:
        return datetime.fromtimestamp(n / 1000.0, tz=timezone.utc).isoformat()
    except (OverflowError, OSError, ValueError):
        return None


def _s(v: Any) -> str:
    return "" if v is None else (v if isinstance(v, str) else str(v))


# --- status -----------------------------------------------------------------

def map_status(raw_status: str) -> str:
    s = re.sub(r"[_-]+", " ", (raw_status or "").lower())
    if "out for delivery" in s:
        return "OUT_FOR_DELIVERY"
    # A failed attempt awaiting re-delivery is still an active, OTP-bearing delivery.
    if re.search(r"unsuccessful|retrying|reattempt|undelivered|delivery failed", s):
        return "OUT_FOR_DELIVERY"
    if "delivered" in s:
        return "DELIVERED"
    if re.search(r"arriving|expected|shipped|on the way|in transit|dispatched", s):
        return "ARRIVING"
    return "OTHER"


def _derive_status(headings: list[str]) -> tuple[str, str]:
    arriving = delivered = False
    for h in headings:
        mapped = map_status(h)
        if mapped == "OUT_FOR_DELIVERY":
            return "OUT_FOR_DELIVERY", h
        arriving = arriving or mapped == "ARRIVING"
        delivered = delivered or mapped == "DELIVERED"
    status = "ARRIVING" if arriving else "DELIVERED" if delivered else "OTHER"
    match = next((h for h in headings if map_status(h) == status), None)
    return status, (match if match is not None else (headings[0] if headings else ""))


# --- product / units --------------------------------------------------------

def _unit_values(order: dict) -> list:
    units = get(order, "units")
    return list(units.values()) if isinstance(units, dict) else []


def _product_title(order: dict, unit: dict) -> str:
    bag = get(order, "productDataBag")
    if not isinstance(bag, dict):
        return ""
    listing_id = _s(get(unit, "metaData.listingId"))
    fsn = _s(get(unit, "metaData.fsn"))
    product = (bag.get(listing_id) if listing_id else None) or (bag.get(fsn) if fsn else None)
    return _s(get(product, "productBasicData.title")).strip()


# --- detail request (which unit to fetch) -----------------------------------

def detail_targets_for(order: dict) -> list[dict]:
    """One detail request per real (non-VAS) unit. The detail endpoint returns that unit's OTP
    plus the order-level address. shareToken is order-level (optional)."""
    order_id = _s(get(order, "orderMetaData.orderId")).strip()
    units = get(order, "units")
    if not order_id or not isinstance(units, dict):
        return []
    share_token = _s(get(order, "accessToOrderDataBag.endUser.id")).strip()
    targets: list[dict] = []
    for unit_id, u in units.items():
        if get(u, "vasItemDetails") is not None:
            continue
        targets.append({"orderId": order_id, "unitId": unit_id, "shareToken": share_token})
    return targets


# --- detail extraction (address + live OTP) ---------------------------------

def extract_detail(detail_json: Any) -> dict:
    """Walk the page-model response for the delivery address object and the active-unit OTP."""
    address = None
    otp = None
    stack = [detail_json]
    while stack:
        n = stack.pop()
        if isinstance(n, dict):
            if address is None:
                a = n.get("address")
                if isinstance(a, dict) and (a.get("addressLine1") or a.get("pinCode")):
                    address = a
            if otp is None:
                v = n.get("otpValue")
                if isinstance(v, str) and v.isdigit() and 4 <= len(v) <= 8:
                    otp = v
            stack.extend(n.values())
        elif isinstance(n, list):
            stack.extend(n)
    return {"address": address, "otp": otp}


def _format_address(addr: Any) -> str:
    if not isinstance(addr, dict):
        return "Address unavailable"
    line1 = _s(addr.get("addressLine1")).strip()
    pin = _s(addr.get("pinCode")).strip()
    if line1:
        return f"{line1} - {pin}" if pin and pin not in line1 else line1
    parts = [_s(addr.get(k)).strip() for k in ("addressLine2", "city", "state")]
    parts.append(pin)
    parts = [p for p in parts if p]
    return ", ".join(parts) if parts else "Address unavailable"


def _unit_otp(unit: dict) -> str | None:
    """OTP from a single unit's otpCallout in the list response (present only while OFD)."""
    oc = get(unit, "deliveryDataBag.otpCallout")
    if not oc:
        return None
    if isinstance(oc, str) and oc.strip():
        return oc.strip()
    if isinstance(oc, dict):
        direct = oc.get("otp") or oc.get("code") or oc.get("value")
        if isinstance(direct, str) and direct.strip():
            return direct.strip()
        m = re.search(r"\b\d{4,8}\b", str(oc))
        if m:
            return m.group(0)
    return None


# --- top-level --------------------------------------------------------------

def orders_array(body: Any) -> list | None:
    arr = get(body, "RESPONSE.multipleOrderDetailsView.orders")
    return arr if isinstance(arr, list) else None


def parse_orders(raw_orders: list, details: dict[str, dict]) -> list[dict]:
    """Explode each order into one row PER UNIT (shipment): a single orderId can yield several
    rows, each with its own item, status, tracking id, and OTP. `details` is keyed
    {orderId: {unitId: {address, otp}}}; the delivery address is order-level (shared)."""
    out: list[dict] = []
    for order in raw_orders:
        if not isinstance(order, dict):
            continue
        order_id = _s(get(order, "orderMetaData.orderId")).strip()
        if not order_id:
            continue
        units = get(order, "units")
        items = list(units.items()) if isinstance(units, dict) else []
        # Skip free add-ons (membership, trust shield); fall back to all units if that leaves nothing.
        real = [(uid, u) for uid, u in items if get(u, "vasItemDetails") is None] or items
        if not real:
            continue

        unit_details = details.get(order_id, {})
        order_address = next((d["address"] for d in unit_details.values() if d.get("address")), None)
        address = _format_address(order_address)
        phone = _s(order_address.get("phoneNumber")).strip() if isinstance(order_address, dict) else ""
        customer = _s(get(order, "accessToOrderDataBag.buyer.name")).strip() or "Unknown customer"
        order_date = get(order, "orderMetaData.orderDate")

        for unit_id, u in real:
            heading = _s(get(u, "metaData.moRedesignHeading")).strip()
            status = map_status(heading)
            delivered = _num(get(u, "deliveryDataBag.promiseDataBag.actualDeliveredDate"))
            promised = _num(get(u, "deliveryDataBag.promiseDataBag.promisedDate"))
            chosen_ms = (delivered or promised) if status == "DELIVERED" else (promised or delivered)
            otp = _unit_otp(u) or unit_details.get(unit_id, {}).get("otp") or None
            if otp:
                # An OTP is only issued for an active delivery -> out for delivery.
                status = "OUT_FOR_DELIVERY"
            out.append({
                "orderId": order_id,
                "trackingId": _s(get(u, "metaData.trackingId")).strip(),
                "customerName": customer,
                "itemName": _product_title(order, u) or "Unknown item",
                "deliveryAddress": address,
                "phone": phone,
                "otp": otp,
                "status": status,
                "rawStatus": heading,
                "activityDateIso": _to_iso(chosen_ms) or _to_iso(order_date) or "",
            })
    return out


def ensure_orders_shape(body: Any) -> list:
    """Raise PARSE_ERROR when the response is not the expected orders shape (likely auth)."""
    if not isinstance(body, dict):
        raise parse_error("Unexpected response shape: not an object (check the session cookie).")
    arr = orders_array(body)
    if arr is None:
        raise parse_error("Unexpected response shape: orders array missing (session likely expired).")
    return arr

File: extension/src/core/parser.ts
// Ported 1:1 from backend/app/parser.py. Parses Flipkart's "My Orders" list + per-order
// detail into the frontend Order shape, exploding each order into one row PER UNIT (shipment).
import type { ParsedOrder, OrderStatus } from "@core/types";
import { parseError } from "./errors";

type Dict = Record<string, unknown>;

// Type guard (preserves narrowing); mirrors Python's `isinstance(x, dict)` — arrays excluded.
function isObj(v: unknown): v is Dict {
  return typeof v === "object" && v !== null && !Array.isArray(v);
}

// --- generic helpers --------------------------------------------------------

export function get(obj: unknown, path: string): unknown {
  let cur: unknown = obj;
  for (const part of path.split(".")) {
    if (!isObj(cur)) return undefined;
    cur = cur[part];
  }
  return cur;
}

// Positive finite number or null (mirrors _num: 0/negative/NaN -> None).
function num(v: unknown): number | null {
  const n = typeof v === "number" ? v : typeof v === "string" ? Number(v) : NaN;
  return Number.isFinite(n) && n > 0 ? n : null;
}

// Epoch-ms -> ISO string, or null when unparseable (mirrors _to_iso).
function toIso(ms: unknown): string | null {
  const n = num(ms);
  if (n === null) return null;
  const d = new Date(n);
  return Number.isNaN(d.getTime()) ? null : d.toISOString();
}

// "" for null/undefined, else the value stringified (mirrors _s).
function str(v: unknown): string {
  return v == null ? "" : typeof v === "string" ? v : String(v);
}

// --- status -----------------------------------------------------------------

export function mapStatus(rawStatus: string): OrderStatus {
  const s = (rawStatus || "").toLowerCase().replace(/[_-]+/g, " ");
  if (s.includes("out for delivery")) return "OUT_FOR_DELIVERY";
  // A failed attempt awaiting re-delivery is still an active, OTP-bearing delivery.
  if (/unsuccessful|retrying|reattempt|undelivered|delivery failed/.test(s)) return "OUT_FOR_DELIVERY";
  if (s.includes("delivered")) return "DELIVERED";
  if (/arriving|expected|shipped|on the way|in transit|dispatched/.test(s)) return "ARRIVING";
  return "OTHER";
}

// --- product / units --------------------------------------------------------

function productTitle(order: unknown, unit: unknown): string {
  const bag = get(order, "productDataBag");
  if (!isObj(bag)) return "";
  const listingId = str(get(unit, "metaData.listingId"));
  const fsn = str(get(unit, "metaData.fsn"));
  const product = (listingId ? bag[listingId] : undefined) ?? (fsn ? bag[fsn] : undefined);
  return str(get(product, "productBasicData.title")).trim();
}

// --- detail request (which unit to fetch) -----------------------------------

export function detailTargetsFor(
  order: unknown,
): { orderId: string; unitId: string; shareToken: string }[] {
  const orderId = str(get(order, "orderMetaData.orderId")).trim();
  const units = get(order, "units");
  if (!orderId || !isObj(units)) return [];
  const shareToken = str(get(order, "accessToOrderDataBag.endUser.id")).trim();
  const targets: { orderId: string; unitId: string; shareToken: string }[] = [];
  for (const [unitId, u] of Object.entries(units)) {
    if (get(u, "vasItemDetails") != null) continue;
    targets.push({ orderId, unitId, shareToken });
  }
  return targets;
}

// --- detail extraction (address + live OTP) ---------------------------------

export function extractDetail(detailJson: unknown): {
  address: Dict | null;
  otp: string | null;
} {
  let address: Dict | null = null;
  let otp: string | null = null;
  const stack: unknown[] = [detailJson];
  while (stack.length) {
    const n = stack.pop();
    if (isObj(n)) {
      if (address === null) {
        const a = n.address;
        if (isObj(a) && (a.addressLine1 || a.pinCode)) address = a;
      }
      if (otp === null) {
        const v = n.otpValue;
        if (typeof v === "string" && /^\d{4,8}$/.test(v)) otp = v;
      }
      for (const val of Object.values(n)) stack.push(val);
    } else if (Array.isArray(n)) {
      for (const val of n) stack.push(val);
    }
  }
  return { address, otp };
}

export function formatAddress(addr: unknown): string {
  if (!isObj(addr)) return "Address unavailable";
  const line1 = str(addr.addressLine1).trim();
  const pin = str(addr.pinCode).trim();
  if (line1) return pin && !line1.includes(pin) ? `${line1} - ${pin}` : line1;
  const parts = ["addressLine2", "city", "state"].map((k) => str(addr[k]).trim());
  parts.push(pin);
  const filtered = parts.filter((p) => p);
  return filtered.length ? filtered.join(", ") : "Address unavailable";
}

// OTP from a single unit's otpCallout in the list response (present only while OFD).
export function unitOtp(unit: unknown): string | null {
  const oc = get(unit, "deliveryDataBag.otpCallout");
  if (!oc) return null;
  if (typeof oc === "string") return oc.trim() || null;
  if (isObj(oc)) {
    const direct = oc.otp ?? oc.code ?? oc.value;
    if (typeof direct === "string" && direct.trim()) return direct.trim();
    const m = JSON.stringify(oc).match(/\b\d{4,8}\b/);
    if (m) return m[0];
  }
  return null;
}

// --- top-level --------------------------------------------------------------

export function ordersArray(body: unknown): unknown[] | null {
  const arr = get(body, "RESPONSE.multipleOrderDetailsView.orders");
  return Array.isArray(arr) ? arr : null;
}

type UnitDetail = { address?: unknown; otp?: string | null };

export function parseOrders(
  rawOrders: unknown[],
  details: Record<string, Record<string, UnitDetail>>,
): ParsedOrder[] {
  const out: ParsedOrder[] = [];
  for (const order of rawOrders) {
    if (!isObj(order)) continue;
    const orderId = str(get(order, "orderMetaData.orderId")).trim();
    if (!orderId) continue;
    const units = get(order, "units");
    const items: [string, unknown][] = isObj(units) ? Object.entries(units) : [];
    // Skip free add-ons (VAS: membership, trust shield); fall back to all units if that empties.
    let real = items.filter(([, u]) => get(u, "vasItemDetails") == null);
    if (real.length === 0) real = items;
    if (real.length === 0) continue;

    const unitDetails = details[orderId] ?? {};
    let orderAddress: unknown = null;
    for (const d of Object.values(unitDetails)) {
      if (d && d.address) {
        orderAddress = d.address;
        break;
      }
    }
    const address = formatAddress(orderAddress);
    const phone = isObj(orderAddress) ? str(orderAddress.phoneNumber).trim() : "";
    const customer = str(get(order, "accessToOrderDataBag.buyer.name")).trim() || "Unknown customer";
    const orderDate = get(order, "orderMetaData.orderDate");

    for (const [unitId, u] of real) {
      const heading = str(get(u, "metaData.moRedesignHeading")).trim();
      let status = mapStatus(heading);
      const delivered = num(get(u, "deliveryDataBag.promiseDataBag.actualDeliveredDate"));
      const promised = num(get(u, "deliveryDataBag.promiseDataBag.promisedDate"));
      const chosenMs = status === "DELIVERED" ? (delivered ?? promised) : (promised ?? delivered);
      const otp = unitOtp(u) || unitDetails[unitId]?.otp || null;
      if (otp) {
        // An OTP is only issued for an active delivery -> out for delivery.
        status = "OUT_FOR_DELIVERY";
      }
      out.push({
        orderId,
        trackingId: str(get(u, "metaData.trackingId")).trim(),
        customerName: customer,
        itemName: productTitle(order, u) || "Unknown item",
        deliveryAddress: address,
        phone,
        otp,
        status,
        rawStatus: heading,
        activityDateIso: toIso(chosenMs) ?? toIso(orderDate) ?? "",
      });
    }
  }
  return out;
}

export function ensureOrdersShape(body: unknown): unknown[] {
  if (!isObj(body)) {
    throw parseError("Unexpected response shape: not an object (check the session cookie).");
  }
  const arr = ordersArray(body);
  if (arr === null) {
    throw parseError("Unexpected response shape: orders array missing (session likely expired).");
  }
  return arr;
}
